### PR TITLE
change type annotation for SharedDataMiddleware to handle common scenarios

### LINKED ISSUE #2958: Type annotations for SharedDataMiddleware produce mypy errors in common circumstances
https://github.com/pallets/werkzeug/blob/5add63c955131fd73531d7369f16b2f1b4e342d4/src/werkzeug/middleware/shared_data.py#L102-L113

`exports` is annotated as `dict[str, str | tuple[str, str]]` (plus the iterable case). Because `dict` is invariant, passing a simple `dict[str, str]` actually produces a mypy error! Users can work around this by manually annotating the data as `dict[str, str | tuple[str, str]]`, but this is very unergonomic.

https://mypy-play.net/?mypy=latest&python=3.12&gist=7194ebea2b603890033453b8d735e232 is a minimal demonstration of this issue, and the recommended workaround: annotating the value as `collections.Mapping` instead.

### PR DESCRIPTION
**Pull request by alex**
fixes #2958

### CODE REVIEW

**Comment by alex** [comment]
Sphinx failure is unrelated I think?

File: src/werkzeug/middleware/shared_data.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 from __future__ import annotations
 
+import collections.abc
```
Comment (davidism) [inline src/werkzeug/middleware/shared_data.py:14]: Alias this to a shorter name with `import collections.abc as cabc`, that's what I've been using elsewhere.

File: src/werkzeug/middleware/shared_data.py [hunk 2]
```diff
@@ -87,6 +88,10 @@ class SharedDataMiddleware:
     :param cache_timeout: the cache timeout in seconds for the headers.
     :param fallback_mimetype: The fallback mimetype for unknown files.
 
+    .. versionchanged:: 3.1.0
```
Comment (davidism) [inline src/werkzeug/middleware/shared_data.py:91]: This can be removed, I haven't been calling out typing fixes with `versionchanged`.

**Comment by alex** [comment]
Ok, I think I rebased it correctly.

**Comment by davidism** [comment]
Looks good, I should be able to work on a new Werkzeug release in the next few days, after I fix this docs issue.

**Comment by alex** [comment]
Sounds great, thanks!

**Comment by alex** [comment]
thanks

**Pull request merged**